src/settings/utils/grouping.js

```diff
@@ -9,6 +9,70 @@ export const getKeyBasedArrayGroup = (array, key) => {
     return acc;
   }, {});
 };
+
+const groupTypeCapabilityByResource = (data) => Object.entries(data).reduce((acc, [type, capabilities]) => {
+  if (!acc[type]) {
+    acc[type] = {};
+  }
+  acc[type] = capabilities.reduce((initial, capability) => {
+    if (!initial[capability.resource]) {
+      initial[capability.resource] = [];
+    }
+    initial[capability.resource].push(capability);
+
+    return initial;
+  }, {});
+
+  return acc;
+}, {});
+
+const groupCapabilitiesObjectByTypeAndResource = (groupedTypeByResource) => {
+  const dataType = [];
+  const proceduralType = [];
+  const settingsType = [];
+
+  Object.entries(groupedTypeByResource).forEach(([type, capabilities]) => {
+    Object.entries(capabilities).forEach(([resource, cap]) => {
+      const actionsObject = cap.reduce((acc, item) => {
+        acc[item.action] = item.id;
+        return acc;
+      }, {});
+      if (type === 'data') {
```

Comment:
Use a switch statement and you can clean this up a little by setting the object to push ahead of time since the only difference is which type it is pushed to.

Reply:
Used object with default fields